folly/SocketPortability.h

```diff
@@ -0,0 +1,148 @@
+  // global scope, then you shouldn't be calling these directly.
+  int accept(int s, struct sockaddr* addr, int* addrlen);
+  int bind(int s, const struct sockaddr* name, int namelen);
+  int connect(int s, const struct sockaddr* name, int namelen);
+  int getpeername(int s, struct sockaddr* name, int* namelen);
+  int getsockname(int s, struct sockaddr* name, int* namelen);
+  int getsockopt(int s, int level, int optname, char* optval, int* optlen);
+  int getsockopt(int s, int level, int optname, void* optval, int* optlen);
+  int inet_aton(const char *cp, struct in_addr *inp);
+  const char* inet_ntop(int af, const void* src, char* dst, socklen_t size);
+  int listen(int s, int backlog);
+  int poll(pollfd* fds, int nfds, int timeout);
+  int recv(int s, char* buf, int len, int flags);
+  int recv(int s, void* buf, int len, int flags);
+  int recvfrom(int s, char* buf, int len, int flags, struct sockaddr* from, int* fromlen);
+  int recvfrom(int s, void* buf, int len, int flags, struct sockaddr* from, int* fromlen);
+  ssize_t recvmsg(int s, struct msghdr* message, int fl);
+  int send(int s, const char* buf, int len, int flags);
+  int send(int s, const void* buf, int len, int flags);
+  ssize_t sendmsg(int s, const struct msghdr *message, int fl);
+  int sendto(int s, const char* buf, int len, int flags, const sockaddr* to, int tolen);
+  int sendto(int s, const void* buf, int len, int flags, const sockaddr* to, int tolen);
+  // We use this first overload to convince MSVC to use our overloads, rather than the
+  // winsock ones, when optval is a const char*.
+  int setsockopt(int s, int level, int optname, const char* optval, int optlen);
+  int setsockopt(int s, int level, int optname, const void* optval, int optlen);
+  int shutdown(int s, int how);
+  int socketpair(int domain, int type, int protocol, int sockot[2]);
+
+  // This is the only function that _must_ be referenced via fsp::
+  // because there is no difference in parameter types to overload
+  // on.
+  int socket(int af, int type, int protocol);
+}}
+
+using namespace folly::socket_portability;
+
+// And a few things that have different names in posix.
+#define sa_family_t ADDRESS_FAMILY
+#undef s_host
+
+#else
+#include <netdb.h>
+#include <poll.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+
+namespace folly { namespace socket_portability {
+  // Unless you have a case where you would normally have
+  // to reference the function as being explicitly in the
+  // global scope, then you shouldn't be calling these directly.
+  inline int bind(int s, const struct sockaddr* name, int namelen) { return ::bind(s, name, namelen); }
+  inline int connect(int s, const struct sockaddr* name, int namelen) { return ::connect(s, name, namelen); }
+  inline int getpeername(int s, struct sockaddr* name, int* namelen) { return ::getpeername(s, name, namelen); }
```

Comment:
Looks like I messed up when I typed up the posix side of the API.... I'll go through and make sure they all get set to socklen_t where they should be.